### PR TITLE
Small fixes to setting Windows file permissions

### PR DESCRIPTION
**Pull request by pitdicker**
On Windows files can have a simple read-only permission flag, and the more complex access rights
and access control lists. The read-only flag is one bit of the file attributes.

To make sure only the read-only flag is toggled and not the other attributes, read the files
attributes and write them back with only this bit changed. By reading and setting the attributes
from the handle the file does not have to be opened twice.

Also directories can not be read-only, here the flag has a different meaning. So we should no
longer report directories as read-only, and also not allow making them read-only.

Note: This pr is split out from the one I am making for #29497. For that I need to be able to
remove the read-only flags from an already open handle. And the read-only detection needed to be
more robust.

### CODE REVIEW

**Comment by pitdicker** [comment]
One more reason we might want this: currently `set_permissions` can be (ab)used to copy file attributes. I think it should just not touch them.

File: src/libstd/sys/windows/fs.rs [hunk 1]
```diff
@@ -333,6 +333,40 @@ impl File {
...
+        match (perm.readonly, attr & c::FILE_ATTRIBUTE_READONLY != 0) {
+            (true, false) =>
+                if attr & c::FILE_ATTRIBUTE_DIRECTORY != 0 {
+                    // this matches directories, dir symlinks, and junctions.
+                    Err(io::Error::new(io::ErrorKind::PermissionDenied,
+                                       "directories can not be read-only"))
+                } else {
+                    self.set_attributes(attr | c::FILE_ATTRIBUTE_READONLY)
+                },
+            (false, true) => self.set_attributes(attr & !c::FILE_ATTRIBUTE_READONLY),
```
Comment (retep998) [inline src/libstd/sys/windows/fs.rs]: Shouldn't this branch also make sure that the file isn't a directory?

Comment (pitdicker) [inline src/libstd/sys/windows/fs.rs]: Oops, yes it should. Thank you!

**Comment by pitdicker** [comment]
I chose to return `Ok(())` if we try to remove the read-only flag from a directory.
This way recursively removing read-only permission is as simple on Windows as it is on Unix, no need to special-case directories.

I still feel a little sad that you can't simply do `chmod -R a-x` on Unix to remove execute permission from all files, but have to special-case directories.

**Comment by pitdicker** [comment]
No high-five today?
r? @alexcrichton? I hope you don't mind

**Comment by pitdicker** [comment]
I messed up git :(. this pr is now included in #31944. So closing this one

**Pull request closed without merging**